gatsby-node.js

```diff
@@ -25,6 +25,36 @@ exports.modifyWebpackConfig = ({config, stage}) => {
   return config;
 };
 
+// Returns an array of duplicate redirect_froms in the Markdown files.
+// The array will be empty if there are no duplicates found.
+// See: https://github.com/reactjs/reactjs.org/issues/1
+function detectDuplicateRedirectFroms(edges) {
+  const redirectFromPaths = {};
+
+  edges.forEach(edge => {
+    const slug = edge.node.fields.slug;
+    if (!edge.node.fields.redirect) {
+      return;
+    }
+    let redirect = JSON.parse(edge.node.fields.redirect);
+    if (!Array.isArray(redirect)) {
+      redirect = [redirect];
+    }
+
+    redirect.forEach(fromPath => {
+      if (!redirectFromPaths.hasOwnProperty(fromPath)) {
+        redirectFromPaths[fromPath] = [];
+      }
+      redirectFromPaths[fromPath].push(slug);
+    });
+  });
+
+  return Object.keys(redirectFromPaths).map(fromPath => ({
+    fromPath,
+    permalinks: redirectFromPaths[fromPath],
+  })).filter(issue => issue.permalinks.length > 1);
+}
```

Comment:
I think I see your point about the separate function being _clearer_ but it feels a bit "meh" to add another for-each.

What are your thoughts about doing the check inside of `createPages`, right before we call `createRedirect`? We could still use this technique of redirect-to-slug-map and just fail the first time we saw a key that already existed. This would only warn about the _first_ duplicate (where as your approach warns about _all_ duplicates) but once this check is in place, there should never be duplicates to begin with, so presumably if we found one to warn about- it would be the only one.

For example:
```diff
diff --git a/content/docs/accessibility.md b/content/docs/accessibility.md
index 77cc52df0..de08b19e5 100644
--- a/content/docs/accessibility.md
+++ b/content/docs/accessibility.md
@@ -2,6 +2,7 @@
 id: accessibility
 title: Accessibility
 permalink: docs/accessibility.html
+redirect_from: "docs/index.html"
 ---
 
 ## Why Accessibility?
diff --git a/gatsby-node.js b/gatsby-node.js
index ab60095a6..07d93da1b 100644
--- a/gatsby-node.js
+++ b/gatsby-node.js
@@ -28,6 +28,9 @@ exports.modifyWebpackConfig = ({config, stage}) => {
 exports.createPages = async ({graphql, boundActionCreators}) => {
   const {createPage, createRedirect} = boundActionCreators;
 
+  // Used to detect and prevent duplicate redirects
+  const redirectToSlugMap = {};
+
   const blogTemplate = resolve('./src/templates/blog.js');
   const communityTemplate = resolve('./src/templates/community.js');
   const docsTemplate = resolve('./src/templates/docs.js');
@@ -114,13 +117,23 @@ exports.createPages = async ({graphql, boundActionCreators}) => {
           redirect = [redirect];
         }
 
-        redirect.forEach(fromPath =>
+        redirect.forEach(fromPath => {
+          if (redirectToSlugMap[fromPath] != null) {
+            console.error(`Duplicate redirect detected from "${fromPath}" to:\n` +
+              `* ${redirectToSlugMap[fromPath]}\n` +
+              `* ${slug}\n`
+            );
+            process.exit(1);
+          }
+
+          redirectToSlugMap[fromPath] = slug;
+
           createRedirect({
             fromPath: `/${fromPath}`,
             redirectInBrowser: true,
             toPath: `/${slug}`,
-          }),
-        );
+          });
+        });
       }
     }
   });
```

Reply:
I get what you mean. I was afraid that it would complicate the `createPage` function too much. I'll change it!